# Графічне та геометричне моделювання

## Лабораторна робота №4. Ділова графіка

**Виконав** Студент групи ІП-32
*Ковальчук Олександр Миронович*

### Завдання

Побудувати таблицю даних згідно варіанту, використовуючи задані математичні формули. За таблицею побудувати графік, гістограму та секторну діаграму.

#### Варіант 12

$$ z = \frac{(x-a)^2}{a} + \frac{(y-b)^2}{b} $$


In [1]:
%matplotlib notebook
import numpy as np
import pandas as pd

/usr/local/lib/python3.5/dist-packages/matplotlib/font_manager.py:273: UserWarning: Matplotlib is building the font cache using fc-list. This may take a moment.
  warnings.warn('Matplotlib is building the font cache using fc-list. This may take a moment.')
/usr/local/lib/python3.5/dist-packages/matplotlib/font_manager.py:273: UserWarning: Matplotlib is building the font cache using fc-list. This may take a moment.
  warnings.warn('Matplotlib is building the font cache using fc-list. This may take a moment.')


Зафіксуємо значення $a$ і $b$:

$$ a = 2 $$
$$ b = 7 $$

In [2]:
a = 2
b = 7

def z(x, y):
    return ((x - a) ** 2) / a + ((y - b) ** 2) / b

Побудуємо таблицю значень:

In [3]:
# generate ranges
xs = list(range(-12, 13, 1))
ys = list(range(-6, 7, 1))

xcnt = len(xs)
ycnt = len(ys)

xs *= ycnt
ys *= xcnt

zs = [z(x, y) for x, y in zip(xs, ys)]

data_dict = {
    'xs': xs,
    'ys': ys,
    'zs': zs,
}

data = pd.DataFrame(data_dict)
data

xs  ys          zs
0   -12  -6  122.142857
1   -11  -5  105.071429
2   -10  -4   89.285714
3    -9  -3   74.785714
4    -8  -2   61.571429
5    -7  -1   49.642857
6    -6   0   39.000000
7    -5   1   29.642857
8    -4   2   21.571429
9    -3   3   14.785714
10   -2   4    9.285714
11   -1   5    5.071429
12    0   6    2.142857
13    1  -6   24.642857
14    2  -5   20.571429
15    3  -4   17.785714
16    4  -3   16.285714
17    5  -2   16.071429
18    6  -1   17.142857
19    7   0   19.500000
20    8   1   23.142857
21    9   2   28.071429
22   10   3   34.285714
23   11   4   41.785714
24   12   5   50.571429
25  -12   6   98.142857
26  -11  -6  108.642857
27  -10  -5   92.571429
28   -9  -4   77.785714
29   -8  -3   64.285714
..   ..  ..         ...
295   8   3   20.285714
296   9   4   25.785714
297  10   5   32.571429
298  11   6   40.642857
299  12  -6   74.142857
300 -12  -5  118.571429
301 -11  -4  101.785714
302 -10  -3   86.285714
303  -9  -2   72.071429
304  -8  -1   59.142857
305  -7   0   47.500000
306  -6   1   37.142857
307  -5   2   28.071429
308  -4   3   20.285714
309  -3   4   13.785714
310  -2   5    8.571429
311  -1   6    4.642857
312   0  -6   26.142857
313   1  -5   21.071429
314   2  -4   17.285714
315   3  -3   14.785714
316   4  -2   13.571429
317   5  -1   13.642857
318   6   0   15.000000
319   7   1   17.642857
320   8   2   21.571429
321   9   3   26.785714
322  10   4   33.285714
323  11   5   41.071429
324  12   6   50.142857

[325 rows x 3 columns]

За таблицею побудуємо графік.

In [4]:
from bokeh.io import output_notebook
output_notebook()

Loading BokehJS ...

In [5]:
from bokeh.charts import Line, show

table = pd.DataFrame(index=xs[:xcnt])
for y_val, df in data.groupby(data.ys):
    table['y = {}'.format(y_val)] = list(df.sort_values(by='xs').zs)

show(Line(table, xlabel='x', ylabel='z(x,y)', title='Залежність z від x'))

Побудуємо карту розподілу (heatmap)

In [6]:
from bokeh.charts import HeatMap, show
show(
    HeatMap(data_dict, x='xs', y='ys', values='zs', stat=None, title='Розподіл z залежно від x та y')
)

Будуємо гістограму розподілу величини $zs$ -- результату функції.

In [7]:
from bokeh.charts import Histogram

hist = Histogram(data, values='zs', title="Розподіл результату функції")
show(hist)

Представимо розподіл результату у вигляді секторної діаграми

In [8]:
from bokeh.charts import Donut

df = data.copy()
df.zs = (df.zs // 15 * 15)
df.zs = df.zs.map(lambda i: '> {}'.format(i))
df.sort_values(by='ys')

donut1 = Donut(df, label=['ys', 'zs'], color='xs', agg='count', title='Розподіл результату функції в розрізі параметра y')
donut2 = Donut(df, label='zs', agg='count', title='Загальний розподіл результату функції')
show(donut1)
show(donut2)

### Висновки

В даній лабораторній роботі були побудовані стовпчаста та секторна діаграма, графік функції та карта розподілу.
В якості даних – бралась функція  

$$ z = \frac{(x-a)^2}{a} + \frac{(y-b)^2}{b} $$ ,де
$$ a = 2, b = 7 $$

Функція була розглянута на проміжку 
$$ x \in [12..12], y \in [6..6] $$

Для заданих інтервалів обчислений результат функції із кроком 1 та на основі цього результату побудовані діаграми розподілу.

При представлені даних в табличному виді маємо $325 \cdot 3 = 975$ окремих елементів для аналізу. При представлені у вигляді графіків, по-перше, збільшується наочність, так залежність результату від змін того чи іншого параметра явно прослідковується, по-друге, кількість окремих елементів для аналізу зменшується до $9 — 11$. Тобто, кількість об'єктів для аналізу зменшилася в $\frac{975}{11} \approx 88$ разів.